LanguageFeatures/Patterns/type_inference_A29_t15.dart

```diff
@@ -0,0 +1,82 @@
+// Copyright (c) 2023, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+/// @assertion
+/// To type check a pattern p being matched against a value of type M:
+/// ...
+/// Map:
+/// i. Calculate the value's entry key type K and value type V, and key context
+///     C:
+///   a. If p has type arguments <K, V> for some K and V then use those, and C
+///     is K.
+///   b. Else if M implements Map<K, V> for some K and V then use those, and C
+///     is K.
+///   c. Else if M is dynamic then K and V are dynamic and C is _.
+///   d. Else K and V are Object? and C is _.
+/// ii. Type-check each key expression using C as the context type.
+/// iii. Type-check each value subpattern using V as the matched value type.
+/// vi. The required type of p is Map<K, V>.
+///
+/// @description Check that each value subpattern is type checked using V as the
+/// matched value type. Test the refutable context and the case when p has no
+/// type arguments <K, V> and M is dynamic
+/// @author sgrekhov22@gmail.com
+
+// SharedOptions=--enable-experiment=patterns
+
+import "../../Utils/expect.dart";
+import "../../Utils/static_type_helper.dart";
+
+String test1() {
+  switch ({"key1": 1 as num} as dynamic) {
+    case {"key1": int  a}:
```

Comment:
The declared type `int` isn't helpful: It just forces `a` to have the declared type `int`, and then we can't see that the type that we're using as the 'matched value type' is `dynamic`.

We won't get an error, even in the case where the matched value type is `String` and the declared type of the pattern variable is `bool`, so I believe that we're learning exactly nothing when `a` has a declared type.

Reply:
Updated